pkg/kstatus/observe/observers/deployment.go

```diff
@@ -0,0 +1,78 @@
+package observers
+
+import (
+	"context"
+
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/cli-utils/pkg/kstatus/observe/common"
+	"sigs.k8s.io/cli-utils/pkg/kstatus/observe/reader"
+	"sigs.k8s.io/cli-utils/pkg/kstatus/status"
+	"sigs.k8s.io/cli-utils/pkg/kstatus/wait"
+)
+
+func NewDeploymentObserver(reader reader.ObserverReader, mapper meta.RESTMapper, rsObserver ResourceObserver) ResourceObserver {
+	return &DeploymentObserver{
+		BaseObserver: BaseObserver{
+			Reader:            reader,
+			Mapper:            mapper,
+			computeStatusFunc: status.Compute,
+		},
+		RsObserver: rsObserver,
+	}
+}
+
+// DeploymentObserver is an observer that can fetch Deployment resources
+// from the cluster, knows how to find any ReplicaSets belonging to the
+// Deployment, and compute status for the deployment.
+type DeploymentObserver struct {
+	BaseObserver
+
+	RsObserver ResourceObserver
+}
+
+func (d *DeploymentObserver) Observe(ctx context.Context, identifier wait.ResourceIdentifier) *common.ObservedResource {
+	deployment, observedResource := d.LookupResource(ctx, identifier)
+	if observedResource != nil {
+		return observedResource
+	}
+	return d.ObserveObject(ctx, deployment)
+}
+
+func (d *DeploymentObserver) ObserveObject(ctx context.Context, deployment *unstructured.Unstructured) *common.ObservedResource {
+	identifier := toIdentifier(deployment)
+
+	observedReplicaSets, err := d.ObserveGeneratedResources(ctx, d.RsObserver, deployment,
+		appsv1.SchemeGroupVersion.WithKind("ReplicaSet").GroupKind(), "spec", "selector")
```

Comment:
This looks up and observe the status for the replicasets, each of which will then look up the pods. This allows the deployment observer to use this information to improve the status (for example, if one or more pods are not Ready, check if they are stuck in some unrecoverable state) and it will also pass references to the generated resources in the event sent back on the go channel. This information is used to show the hierarchy of resources in the screen shot in the PR description.
I have considered looking at the owner references too. Although I think in practice they should always match since I think the controllers will "adopt" resources that matches the selector. But need to verify that.
When you are mentioning current generation here, do you mean the newest replicaset? I agree those are most useful, but the old ones can be interesting to visualize how pods are updated during the rollout.

Reply:
>  Although I think in practice they should always match since I think the controllers will "adopt" resources that matches the selector. But need to verify that.

IIRC the case where that won't be true is overlapping selectors: if a resource is already owned, it will not be adopted.